### PR TITLE
Internal: DevTools - A JS Deprecation function that calculates soft/hard versions. [ED-6897]

### PR DESCRIPTION
**Pull request by iNewLegend**

### CODE REVIEW

File: modules/dev-tools/assets/js/editor/dev-tools.js [hunk 1]
```diff
@@ -1,14 +1,102 @@
...
 	}
+
+	/**
+	 * @param {string} version
+	 *
+	 * @return {Version}
+	 */
+	parseVersion( version ) {
+		const versionParts = version.split( '.' ),
+			versionPartsLength = versionParts.length;
```
Comment (StyleShit) [inline modules/dev-tools/assets/js/editor/dev-tools.js:25]: It's almost the exact same length of a string 😅 
I think it's redundant
```
versionPartsLength
versionParts.length
```

File: modules/dev-tools/assets/js/editor/dev-tools.js [hunk 2]
```diff
@@ -1,14 +1,102 @@
...
+
+		if ( versionPartsLength < 3 || versionPartsLength > 4 ) {
+			throw new RangeError( 'Invalid Semantic Version string provided' );
+		}
+
+		const [ major1, major2, minor ] = versionParts,
```
Comment (StyleShit) [inline modules/dev-tools/assets/js/editor/dev-tools.js:31]: I think it should be:
```suggestion
		const [ major, minor, patch ] = versionParts,
```

File: modules/dev-tools/assets/js/editor/dev-tools.js [hunk 3]
```diff
@@ -1,14 +1,102 @@
...
+			result = {
+				major1: parseInt( major1 ),
+				major2: parseInt( major2 ),
+				minor: parseInt( minor ),
+			};
+
+		if ( versionPartsLength > 3 ) {
+			result.build = versionParts[ 3 ];
+		}
```
Comment (StyleShit) [inline modules/dev-tools/assets/js/editor/dev-tools.js:40]: Maybe deconstruct the `build` in line 31, and if it's defined just add that to the result object upon creation?

```suggestion
			result = {
				major1: parseInt( major1 ),
				major2: parseInt( major2 ),
				minor: parseInt( minor ),
				...( build ? { build } : {} ),
			};
```

Oe maybe:
```suggestion
			result = {
				major1: parseInt( major1 ),
				major2: parseInt( major2 ),
				minor: parseInt( minor ),
				build,
			};
```

This way you can avoid using a variable and just return the result
```
return {
	major1: parseInt( major1 ),
	major2: parseInt( major2 ),
	minor: parseInt( minor ),
	build,
}
```

File: modules/dev-tools/assets/js/editor/dev-tools.js [hunk 4]
```diff
@@ -1,14 +1,102 @@
...
+	getTotalMajor( versionObj ) {
+		let total = parseInt( `${ versionObj.major1 }${ versionObj.major2 }0` );
+
+		total = Number( ( total / 10 ).toFixed( 0 ) );
+
+		if ( versionObj.major2 > 9 ) {
+			total = versionObj.major2 - 9;
+		}
+
+		return total;
```
Comment (StyleShit) [inline modules/dev-tools/assets/js/editor/dev-tools.js:59]: explain

File: modules/dev-tools/assets/js/editor/dev-tools.js [hunk 5]
```diff
@@ -1,14 +1,102 @@
...
+	 * @param {string} version2
+	 *
+	 * @return {number}
+	 */
+	compareVersion( version1, version2 ) {
+		const versionObj1 = this.parseVersion( version1 ),
+			versionObj2 = this.parseVersion( version2 );
+
+		[ versionObj1, versionObj2 ]
+			.forEach( ( version ) => version.total = this.getTotalMajor( version ) );
```
Comment (StyleShit) [inline modules/dev-tools/assets/js/editor/dev-tools.js:73]: maybe just use map and reduce?
```suggestion
		return [ versionObj1, versionObj2 ]
			.map( this.getTotalMajor )
			.reduce( ( total1, total2 ) => total1 - total2 );
```

File: tests/qunit/tests/modules/dev-tools/assets/js/editor/dev-tools.spec.js [hunk 6]
```diff
@@ -9,8 +15,51 @@ QUnit.module( 'File: modules/dev-tools/assets/js/editor/dev-tools.js', () => {
...
 
-		elementor.devTools.notifyDeprecated();
+		elementor.devTools.notifyBackendDeprecations();
 
 		elementorCommon.helpers.softDeprecated = softDeprecatedOrig;
 	} );
+
+	QUnit.test( 'parseVersion()', ( assert ) => {
+		// Act.
+		const version = elementor.devTools.parseVersion( '1.2.3.beta' );
```
Comment (StyleShit) [inline tests/qunit/tests/modules/dev-tools/assets/js/editor/dev-tools.spec.js:25]: what about versions without build?

File: tests/qunit/tests/modules/dev-tools/assets/js/editor/dev-tools.spec.js [hunk 7]
```diff
@@ -9,8 +15,51 @@ QUnit.module( 'File: modules/dev-tools/assets/js/editor/dev-tools.js', () => {
...
+			build: 'beta',
+			major1: 1,
+			major2: 2,
+			minor: 3,
+		} );
+	} );
+
+	QUnit.test( 'compareVersion()', ( assert ) => {
+		// Act.
+		const result = elementor.devTools.compareVersion( '1.2.3.beta', '1.2.3.beta' );
```
Comment (StyleShit) [inline tests/qunit/tests/modules/dev-tools/assets/js/editor/dev-tools.spec.js:38]: what about greater and smaller versions?

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: modules/dev-tools/assets/js/editor/dev-tools.js (hunk 1 above)
Comment (iNewLegend) [reply modules/dev-tools/assets/js/editor/dev-tools.js]: Handled.

File: modules/dev-tools/assets/js/editor/dev-tools.js (hunk 2 above)
Comment (iNewLegend) [reply modules/dev-tools/assets/js/editor/dev-tools.js]: In WordPress world it `major1, major2, minor`.

File: modules/dev-tools/assets/js/editor/dev-tools.js (hunk 3 above)
Comment (iNewLegend) [reply modules/dev-tools/assets/js/editor/dev-tools.js]: Handled.

File: modules/dev-tools/assets/js/editor/dev-tools.js (hunk 5 above)
Comment (iNewLegend) [reply modules/dev-tools/assets/js/editor/dev-tools.js]: Handled

File: tests/qunit/tests/modules/dev-tools/assets/js/editor/dev-tools.spec.js (hunk 6 above)
Comment (iNewLegend) [reply tests/qunit/tests/modules/dev-tools/assets/js/editor/dev-tools.spec.js]: Handled

File: tests/qunit/tests/modules/dev-tools/assets/js/editor/dev-tools.spec.js (hunk 7 above)
Comment (iNewLegend) [reply tests/qunit/tests/modules/dev-tools/assets/js/editor/dev-tools.spec.js]: Handled

File: modules/dev-tools/assets/js/editor/dev-tools.js (hunk 4 above)
Comment (iNewLegend) [reply modules/dev-tools/assets/js/editor/dev-tools.js]: Handled.

File: modules/dev-tools/assets/js/editor/deprecation.js [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+		}
+
+		const [ major1, major2, minor, build = '' ] = versionParts;
+
+		return {
+				major1: parseInt( major1 ),
+				major2: parseInt( major2 ),
+				minor: parseInt( minor ),
+				build,
+			};
```
Comment (StyleShit) [inline modules/dev-tools/assets/js/editor/deprecation.js:29]:
```suggestion
		return {
			major1: parseInt( major1 ),
			major2: parseInt( major2 ),
			minor: parseInt( minor ),
			build,
		};
```

**Review by StyleShit — APPROVED** [review]

**Review by matipojo — APPROVED** [review]
Nice!!

**Pull request merged**